### ISSUE TITLE
Reassigning the memory free pointer does not cause an error.

### ISSUE THREAD

**Issue by Subway2023** [matter-labs/era-compiler-solidity#272]
```solidity
contract TestMStore {
    function testFunction() public pure returns (uint256) {
        uint256 result;
        assembly {
            let value := 42
            mstore(0x40, value)
            result := mload(0x40)
        }
        return result;
    }
}
```
solc+evm: execution error
solc(viaIR)+evm: execution is normal
zksolc+era_test_node: execution is normal

I believe this difference is due to zksolc compiling from Yul. Therefore, this difference needs to be reflected in the [documentation](https://docs.zksync.io/zksync-protocol/differences/evm-instructions).

**Comment by hedgar2017** [comment]
@Subway2023 yes, Yul is still our default codegen. It is mentioned in zksolc docs here: https://matter-labs.github.io/era-compiler-solidity/latest/02-command-line-interface.html#--codegen

Let me see. Perhaps, we can finally switch our default codegen to EVM assembly.

**Comment by hedgar2017** [comment]
@Subway2023 the recent versions of `zksolc` produce a warning for codegen:
```
Warning: The `codegen` setting will become mandatory in future versions of zksolc. Please set it to either `evmla` or `yul`
```
It is not verbose enough and doesn't explain what is our default and why.

We will make it more informative and keep it for a few months before converting to a hard error.
